File: go/src/di_string_match/942.di_string_match.go
// 2018-11-19 16:14
package main

import "fmt"

func diStringMatch(S string) []int {
	ret := []int{}
	//   I D I D
	// 0 4 1 3 2
	lo, hi := 0, len(S)
	for i := 0; i < len(S); i++ {
		if S[i] == 'I' {
			ret = append(ret, lo)
			lo++
		} else {
			ret = append(ret, hi)
			hi--
		}
	}
	ret = append(ret, lo)
	return ret
}

func main() {
	fmt.Println(diStringMatch("DDDII"))
}

File: rust/di_string_match/942_di_string_match.rs
// 2018-12-26 14:50
// @lc id=942 lang=rust
struct Solution {}
impl Solution {
    pub fn di_string_match(s: String) -> Vec<i32> {
        let (mut lo, mut hi) = (0 as i32, s.len() as i32);
        let mut ret: Vec<_> = vec![];
        s.chars().for_each(|c| match c {
            'I' => {
                ret.push(lo);
                lo += 1;
            }
            'D' => {
                ret.push(hi);
                hi -= 1;
            }
            _ => (),
        });
        ret.push(lo);
        return ret;
    }
}

fn main() {
    println!("{:?}", Solution::di_string_match("IDID".to_owned()));
}
